Augment NullAway error / fix serialization for Annotator auto fix mode

nimakarimipour opened the pull request:
This PR extends NullAway's error serialization to include additional metadata needed by the Annotator for automatic code-fix generation. 
A new serialization `v4` adds origin tracing, when a local variable triggers an error, `OriginScanner` walks the enclosing method's AST to find the original nullable symbols (fields, parameters, method calls) that contributed to its value, telling the Annotator what to annotate. 
For dereference errors, a typed `NullableExpressionInfo` record captures structured metadata about the nullable expression (source text, symbol kind, enclosing class, annotation status, symbol name, position). 
Since an error can have 0–N origin traces each with nested location data, `v4` switches the error output from flat `TSV` to `XML` (`errors.xml`) to naturally represent this hierarchical, variable-length structure; earlier versions remain `TSV`. 
The ErrorBuilder API provides simple overloads for call sites that don't need origin tracing, while `WithInfo` variants are reserved for dereference error sites. `MayBeNullableInquiry` is now private to NullAway. Three new tests validate XML output for field dereferences, local variable origin traces, and origin traces through enhanced-for loops.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Enhanced fix-serialization to include richer nullable-expression metadata (plus origin/provenance details) for more nullability diagnostics.
  * Added XML-based reporting output in the latest serialization version, including `errors.xml` with additional serialized elements.
* **Bug Fixes**
  * Improved consistency and completeness of serialized error context, especially for dereference/unboxing and chained-origin scenarios.
* **Tests**
  * Expanded XML serialization coverage and updated assertions/wording to match the new diagnostic and serialized output behavior.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

nimakarimipour commented:
@coderabbitai make a comparison between yours and Gitar.ai. Which one is more usefull ?

nimakarimipour committed:
Augment NullAway error / fix serialization for Annotator auto fix mode

Extend the fix-serialization output with extra metadata needed by
Annotator's automatic code-fix mode:

- Add serialization version 4 with a new `infos` column.
- Emit errors.xml alongside errors.tsv with per-error XML records,
  including origin traces for @Nullable local variables (walking
  assignments / initializers back to the fields, parameters, or method
  returns that feed them).
- Thread a Map<String, String> of site-specific metadata through
  ErrorBuilder for dereference / enhanced-for errors.
- Thread nullable-expression + a MayBeNullableInquiry callback through
  every createErrorDescription call site (NullAway, GenericsChecks,
  contract handlers, field-contract handlers) so the origin scanner can
  run during reporting.

msridhar commented on build.gradle:
```diff
@@ -80,7 +80,7 @@ subprojects { project ->
             "-Xlint:rawtypes",
             "-Xlint:unchecked",
-            "-Werror",
+            //            "-Werror",
```
yes @nimakarimipour we should undo this change and fix any issues

nimakarimipour replied:
@msridhar Sorry I completely forgot that I added this to speed up local testing :) , Added `-Werror` back.

nimakarimipour committed:
Address CodeRabbit feedback on serialization changes

Restore -Werror and fix surfaced warnings. Serialize the infos column
in V4 TSV rows so the header and row widths match. Use Locale.ROOT for
machine-readable kind metadata. Preserve trailing empties when splitting
tab-separated location strings. Remove a stray System.out.println from
the compiler path.

nimakarimipour commented on nullaway/src/main/java/com/uber/nullaway/fixserialization/Serializer.java:
```diff
@@ -195,4 +203,25 @@ public static String serializeSymbol(@Nullable Symbol symbol, SerializationAdapt
+        default -> sb.append(c);
+      }
+    }
+  }
```
Not worth it, maybe follow up.

nimakarimipour commented on nullaway/src/main/java/com/uber/nullaway/fixserialization/Serializer.java:
```diff
@@ -73,6 +76,10 @@ public Serializer(FixSerializationConfig config, SerializationAdapter serializat
   public void serializeErrorInfo(ErrorInfo errorInfo) {
     errorInfo.initEnclosing();
     appendToFile(serializationAdapter.serializeError(errorInfo), errorOutputPath);
+    // append to xml as well.
+    StringBuilder xml = new StringBuilder();
+    errorInfo.appendXml(xml, serializationAdapter);
+    appendToFile(xml.toString(), errorOutputXmlPath);
   }
```
Not worth it, this is not a hot path.

nimakarimipour commented on nullaway/src/main/java/com/uber/nullaway/fixserialization/out/ErrorInfo.java:
```diff
@@ -134,4 +156,70 @@ public int getOffset() {
+    // check if the path to the symbol exists
+    Symbol.ClassSymbol enclosingClass = castToNonNull(ASTHelpers.enclosingClass(symbol));
+    URI pathInURI = enclosingClass.sourcefile != null ? enclosingClass.sourcefile.toUri() : null;
+    return Serializer.pathToSourceFileFromURI(pathInURI) != null;
```
I already added a TODO, will address this in follow up PR.

nimakarimipour commented on nullaway/src/main/java/com/uber/nullaway/NullAway.java:
```diff
@@ -1944,12 +1979,24 @@ public Description matchEnhancedForLoop(EnhancedForLoopTree tree, VisitorState s
+          Boolean.toString(
+              !codeAnnotationInfo.isSymbolUnannotated(derefedSymbol, config, handler)));
+      args.put("symbol", Serializer.serializeSymbol(derefedSymbol, adapter));
+      args.put("position", Integer.toString(((JCTree) expr).pos().getStartPosition()));
```
Fixed

nimakarimipour commented on nullaway/src/main/java/com/uber/nullaway/fixserialization/scanners/OriginTrace.java:
```diff
@@ -0,0 +1,65 @@
+  public OriginTrace(Symbol origin, Tree trace) {
+    this.origin = origin;
+    this.trace = trace;
+  }
```
Fixed

nimakarimipour committed:
Convert OriginTrace to a record

Removes the boilerplate constructor, getters, equals, and hashCode in
favor of a Java 17 record. Call sites switch from getOrigin()/getTrace()
to the generated origin()/trace() accessors.

nimakarimipour committed:
Address second round of CodeRabbit feedback

- V4 infos column now uses an explicit key=value,... codec with
  backslash escaping for \, =, and , so values can round-trip without
  being confused with entry separators. Previously relied on
  Map.toString() which produced ambiguous output.
- ExpressionToSymbolScanner gains a visitMethodInvocation override so
  nullable method-return origins (e.g. foo() in Object f = foo()) are
  captured instead of dropped.
- Serializer now gates errors.xml creation and appends behind
  serialization version >= 4, so V1/V3 users don't get a surprise
  artifact. Documented the fragment-stream contract.
- OriginScanner accepts a diagnostic start offset and skips
  assignments, variable declarations, and for-each loops that start at
  or after it. ErrorBuilder passes the error tree position so origins
  from non-reaching future assignments are no longer serialized.

nimakarimipour committed:
Address CodeRabbit comments

nimakarimipour committed:
Merge branch 'master' into nimak/autocodefix

nimakarimipour committed:
dump TSV and use XML

nimakarimipour committed:
Merge branch 'nimak/autocodefix' of github.com:nimakarimipour/NullAway into nimak/autocodefix

nimakarimipour committed:
Use StAX XMLStreamWriter for error XML serialization

Replace hand-coded string concatenation and manual escape table with
javax.xml.stream.XMLStreamWriter (JDK built-in StAX). ErrorInfo,
AbstractSymbolLocation, and SymbolLocation now write to an
XMLStreamWriter; Serializer wraps a per-error StringWriter and appends
the result to errors.xml. The on-disk fragment-stream layout is
unchanged.

msridhar committed:
Merge branch 'master' into nimak/autocodefix

msridhar commented:
@nimakarimipour can you update the PR summary to better describe the changes overall?  I don't really understand what are the key new functionalities here, and why we need to switch from TSV to XML.  Also, it seems like code coverage is pretty low for this patch (see the codecov comment).  Can we bring that up with more tests?

msridhar commented on nullaway/src/main/java/com/uber/nullaway/fixserialization/out/ErrorInfo.java:
```diff
@@ -59,8 +71,18 @@ public class ErrorInfo {
   private final @Nullable Path path;
 
+  /** Extra argument regarding the error required to generate a fix automatically. */
+  private final Map<String, String> infos;
```
Do we know the shape of this map at all, like what keys are expected?

nimakarimipour replied:
Yes, replaced the map with a typed NullableExpressionInfo record with explicit fields.

msridhar commented on nullaway/src/main/java/com/uber/nullaway/fixserialization/scanners/ExpressionToSymbolScanner.java:
```diff
@@ -0,0 +1,141 @@
+
+  @Override
+  public Set<Symbol> reduce(Set<Symbol> r1, Set<Symbol> r2) {
+    if (r2 == null && r1 == null) {
```
I see a few null checks on symbols not marked as `@Nullable` like these here.  Given that all our code is checked with NullAway, these should not be necessary, correct?

nimakarimipour replied:
Added @Nullable to the parameters to make this explicit. Other impl of method in scanner pass `null` to reduce.

msridhar replied:
Ok.  The `:nullaway:buildWithNullAway` task checks if the code compiles with the latest NullAway.  Seems like that is failing on CI, so we still have some issues like this.

nimakarimipour replied:
Fixed in latest commit. NullAway is passing now.

nimakarimipour committed:
remove map

nimakarimipour committed:
Merge branch 'nimak/autocodefix' of github.com:nimakarimipour/NullAway into nimak/autocodefix

nimakarimipour committed:
simplify

nimakarimipour committed:
add nullable annotations

nimakarimipour committed:
Merge branch 'master' into nimak/autocodefix

nimakarimipour committed:
add more tests

nimakarimipour committed:
Merge branch 'nimak/autocodefix' of github.com:nimakarimipour/NullAway into nimak/autocodefix

nimakarimipour committed:
simplify

msridhar commented on nullaway/src/main/java/com/uber/nullaway/fixserialization/out/ErrorInfo.java:
```diff
@@ -130,8 +153,82 @@ public int getOffset() {
+   * @param symbol The symbol to check.
+   * @return True if the symbol is from annotated code, false otherwise.
+   */
+  private static boolean isAnnotated(Symbol symbol) {
```
We already have the `com.uber.nullaway.CodeAnnotationInfo` class for querying whether symbols are annotated.  There is a field `codeAnnotationInfo` in `NullAway` holding the object.  Is it hard to pass down to here?

nimakarimipour replied:
Fixed, passing `codeAnnotationInfo` now.

msridhar commented on nullaway/src/main/java/com/uber/nullaway/fixserialization/out/NullableExpressionInfo.java:
```diff
@@ -0,0 +1,58 @@
+    int position) {
+
+  public void writeXml(XMLStreamWriter writer) throws XMLStreamException {
+    writer.writeStartElement("infos");
```
Is `infos` really a good name?  Maybe `nullableExpressionInfo` instead?

nimakarimipour replied:
Sure, Renamed to `nullableExpressionInfo`.

msridhar commented on nullaway/src/main/java/com/uber/nullaway/fixserialization/scanners/ExpressionToSymbolScanner.java:
```diff
@@ -0,0 +1,143 @@
+import org.jspecify.annotations.Nullable;
+
+/** Scanner that finds the symbols of all identifiers in expressions. */
+public class ExpressionToSymbolScanner
```
Let's document the `BiPredicate` type argument.  Looks like it is some kind of nullability test?

nimakarimipour replied:
Yes Added the documentation.

msridhar commented on nullaway/src/main/java/com/uber/nullaway/fixserialization/scanners/ExpressionToSymbolScanner.java:
```diff
@@ -0,0 +1,143 @@
+      return Set.of();
+    }
+    Symbol symbol = defaultResult(node);
+    return symbol == null ? Set.of() : Set.of(symbol);
```
This pattern appears multiple times identically, can we extract into a separate method and reuse?

nimakarimipour replied:
Extracted and resued method `symbolIfNullable()` in all 4 places.

msridhar commented on nullaway/src/main/java/com/uber/nullaway/fixserialization/scanners/OriginScanner.java:
```diff
@@ -0,0 +1,208 @@
+  }
+
+  @Override
+  public Set<OriginTrace> reduce(@Nullable Set<OriginTrace> r1, @Nullable Set<OriginTrace> r2) {
```
Is this just computing a set union?  If so can we use [the Guava method](https://guava.dev/releases/22.0/api/docs/com/google/common/collect/Sets.html#union-java.util.Set-java.util.Set-)?

nimakarimipour replied:
Reused `Sets.union`.

msridhar commented on nullaway/src/main/java/com/uber/nullaway/fixserialization/scanners/OriginTrace.java:
```diff
@@ -0,0 +1,36 @@
+ * @param origin The origin symbol.
+ * @param trace The tree where the origin contributes to the value of the local variable.
+ */
+public record OriginTrace(Symbol origin, Tree trace) {}
```
When I think of a trace I think of a sequence of things, but this is just a `Tree`.  Maybe `OriginLocation`?

nimakarimipour replied:
Replaced it with `OriginLocation`.

msridhar commented on nullaway/src/main/java/com/uber/nullaway/NullAway.java:
```diff
@@ -1944,12 +1974,28 @@ public Description matchEnhancedForLoop(EnhancedForLoopTree tree, VisitorState s
-            MessageTypes.DEREFERENCE_NULLABLE,
-            "enhanced-for expression " + state.getSourceForNode(expr) + " is @Nullable");
+    Symbol derefedSymbol = ASTHelpers.getSymbol(expr);
+    NullableExpressionInfo exprInfo =
```
This code should only run if serialization is enabled, correct?  Let's be careful to check for the serialization flag before running any code as part of the main analysis, as much as possible

nimakarimipour replied:
Sure, This is only built when serialization is active.

msridhar reviewed (commented):
Some more comments.  More broadly, inside `NullAway`, please ensure as much as possible that logic related to serialization only runs when the relevant config flag is enabled, to minimize risk to the main NullAway paths.

nimakarimipour committed:
Address review feedback on fix serialization scanners

- Use CodeAnnotationInfo to determine origin symbol annotatedness instead
  of the path-existence heuristic; thread handler/config/codeAnnotationInfo
  through ErrorBuilder and SerializationService into ErrorInfo
- Rename NullableExpressionInfo XML element "infos" -> "nullableExpressionInfo"
- Fix buildWithNullAway failures: skip building NullableExpressionInfo when
  the expression has no source text (getSourceForNode is @Nullable)
- Document the BiPredicate nullability-test parameter of ExpressionToSymbolScanner
- Extract the repeated visit* body into symbolIfNullable()
- Use Guava Sets.union in the reduce() methods
- Rename OriginTrace -> OriginLocation (and its trace component -> tree)
- Only build NullableExpressionInfo when serialization is active, via a
  shared buildNullableExpressionInfo() helper

nimakarimipour committed:
Merge branch 'master' into nimak/autocodefix

nimakarimipour commented:
@msridhar Thank you for the review, addressed all comments, this is ready for another round of review.

msridhar committed:
Merge branch 'master' into nimak/autocodefix

msridhar commented on nullaway/src/main/java/com/uber/nullaway/NullAway.java:
```diff
@@ -198,6 +202,11 @@ public class NullAway extends BugChecker
       ImmutableSet.of(TYPE_USE, TYPE_PARAMETER);
 
+  /** Returns {@code true} if the given expression may be null. */
+  private boolean mayBeNullInquiry(ExpressionTree tree, VisitorState state) {
```
Do we really need this wrapper method?

msridhar commented on nullaway/src/main/java/com/uber/nullaway/NullAway.java:
```diff
@@ -298,9 +309,10 @@ public GenericsChecks getGenericsChecks() {
+    errorBuilder = new ErrorBuilder(config, "", ImmutableSet.of(), handler);
     // annoying to leak `this` here; we assign the field last to make it as safe as possible
     genericsChecks = new GenericsChecks(this, config, handler);
+    this.adapter = SerializationAdapter.getAdapterForVersion(SerializationAdapter.LATEST_VERSION);
```
This should be before line 313; see the comment on that line

msridhar commented on nullaway/src/main/java/com/uber/nullaway/NullAway.java:
```diff
@@ -314,9 +326,10 @@ public NullAway(ErrorProneFlags flags) {
+    errorBuilder = new ErrorBuilder(config, canonicalName(), allSuppressionNames, handler);
     // annoying to leak `this` here; we assign the field last to make it as safe as possible
     genericsChecks = new GenericsChecks(this, config, handler);
+    this.adapter = SerializationAdapter.getAdapterForVersion(SerializationAdapter.LATEST_VERSION);
```
This should be before line 330; see the comment on that line

msridhar commented on nullaway/src/main/java/com/uber/nullaway/NullAway.java:
```diff
@@ -289,6 +298,8 @@ public GenericsChecks getGenericsChecks() {
     return genericsChecks;
   }
 
+  private final SerializationAdapter adapter;
```
Can we Javadoc this field?

msridhar commented on nullaway/src/main/java/com/uber/nullaway/NullAway.java:
```diff
@@ -1959,12 +1989,19 @@ public Description matchEnhancedForLoop(EnhancedForLoopTree tree, VisitorState s
+    Symbol derefedSymbol = ASTHelpers.getSymbol(expr);
+    NullableExpressionInfo exprInfo = buildNullableExpressionInfo(expr, derefedSymbol, state);
+    String message = "enhanced-for expression '" + state.getSourceForNode(expr) + "' is @Nullable";
+    ErrorMessage errorMessage = new ErrorMessage(MessageTypes.DEREFERENCE_NULLABLE, message);
```
I'm not sure why we were constructing the error message outside the `mayBeNullExpr` condition check below.  Can we move this code under the `if`?

msridhar commented on nullaway/src/main/java/com/uber/nullaway/NullAway.java:
```diff
@@ -2034,10 +2071,18 @@ private void doUnboxingCheck(VisitorState state, ExpressionTree... expressions)
-              new ErrorMessage(MessageTypes.UNBOX_NULLABLE, "unboxing of a @Nullable value");
+              new ErrorMessage(
+                  MessageTypes.UNBOX_NULLABLE,
+                  "unboxing of a @Nullable value - " + state.getSourceForNode(tree));
```
Since we're modifying here can we do the following instead?

```suggestion
                  "unboxing of a @Nullable expression '" + state.getSourceForNode(tree)) + "'";
```

We're trying to consistently enclose expressions in single quotes now.  And, can we update relevant test error message expectations for this change?

msridhar commented on nullaway/src/main/java/com/uber/nullaway/ErrorBuilder.java:
```diff
@@ -99,27 +110,94 @@ public Description createErrorDescription(
+        null);
+  }
+
+  public Description createErrorDescription(
```
Javadoc please

msridhar commented on nullaway/src/main/java/com/uber/nullaway/ErrorBuilder.java:
```diff
@@ -99,27 +110,94 @@ public Description createErrorDescription(
+        null);
+  }
+
+  public Description createErrorDescriptionWithInfo(
```
Javadoc

msridhar commented on nullaway/src/main/java/com/uber/nullaway/ErrorBuilder.java:
```diff
@@ -99,27 +110,94 @@ public Description createErrorDescription(
+        nullableExpressionInfo);
   }
 
-  /**
-   * create an error description for a nullability warning
-   *
-   * @param errorMessage the error message object.
-   * @param suggestTree the location at which a fix suggestion should be made
-   * @param descriptionBuilder the description builder for the error.
-   * @param state the visitor state (used for e.g. suppression finding).
-   * @param nonNullTarget if non-null, this error involved a pseudo-assignment of a @Nullable
-   *     expression into a @NonNull target, and this parameter is the Symbol for that target.
-   * @return the error description
-   */
```
Why is this Javadoc deleted?

msridhar commented on nullaway/src/main/java/com/uber/nullaway/ErrorBuilder.java:
```diff
@@ -99,27 +110,94 @@ public Description createErrorDescription(
+        errorMessage, suggestTree, descriptionBuilder, state, null, nonNullTarget, null, null);
+  }
+
+  public Description createErrorDescription(
```
Javadoc.  Also, can you explain why we need so many overloads?  Maybe as a follow-up to this PR we should introduce some kind of `Builder` type for these error descriptions?

msridhar replied:
I'm still curious why we need all these overloads.  Maybe a coding agent could find a way to replace them with some kind of builder API?

msridhar commented on nullaway/src/main/java/com/uber/nullaway/ErrorBuilder.java:
```diff
@@ -99,27 +110,94 @@ public Description createErrorDescription(
+        null);
+  }
+
+  public Description createErrorDescriptionWithInfo(
```
Javadoc

msridhar reviewed (commented):
A few more comments

nimakarimipour committed:
Address review feedback: javadoc error-description overloads, reorder adapter init, tidy unboxing/enhanced-for error construction

msridhar committed:
Merge branch 'master' into nimak/autocodefix

msridhar committed:
fix expected error

nimakarimipour committed:
Remove mayBeNullExpr wrapper by reordering BiPredicate type parameters

nimakarimipour committed:
Fix Verify crash in origin tracing by making nullability inquiry path-aware

OriginScanner walked the enclosing method but evaluated the nullability
inquiry (mayBeNullExpr) against the fixed diagnostic path. For a traced
expression elsewhere in the method, path-sensitive checks looked at the
wrong program point, and in JSpecify mode a generic-return method call as
an origin tripped Verify.verify in mayBeNullMethodCall. Route every
inquiry through a state whose path locates the visited expression.

nimakarimipour commented:
Thanks for the review, this is ready for another review @msridhar. I also made the PR on Annotator which works with this version of NullAway https://github.com/ucr-riple/NullAwayAnnotator/pull/277

msridhar committed:
Merge branch 'master' into nimak/autocodefix

msridhar committed:
cleanup

msridhar committed:
fix nullaway warning

msridhar commented:
@nimakarimipour what do you think of the changes I made?  The changes to `NullAway` are more minimal now, and more of the logic lives in `ErrorBuilder`, which I think makes more sense.

nimakarimipour committed:
Merge branch 'master' into nimak/autocodefix

nimakarimipour commented:
@msridhar Yes these changes looks good to me. This is ready to be landed.
